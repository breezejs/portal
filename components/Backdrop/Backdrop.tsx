import React, {Fragment, memo, ReactNode} from 'react';
import styled from 'styled-components';
import {GlobalStylesTheme} from '../../core/GlobalStyles';

const Overlay = styled.div`
  background-color: ${GlobalStylesTheme.black};
  bottom: 0;
  left: 0;
  opacity: .5;
  position: fixed;
  right: 0;
  top: 0;
  z-index: ${GlobalStylesTheme.zIndexBackdrop};
`;

interface IBackdropProps {
  children?: ReactNode;
  show: boolean;
}

function Backdrop ({children, show}: IBackdropProps) {
  function renderBackdrop () {
    return(
      <Overlay>{children}</Overlay>
    );
  }

  return(
    <Fragment>{show && renderBackdrop()}</Fragment>
  );
}

export default memo(Backdrop);
