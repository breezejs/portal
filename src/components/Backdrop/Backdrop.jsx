import PropTypes from 'prop-types';
import React, {Fragment, memo} from 'react';
import styled from 'styled-components';
import {GlobalStylesTheme} from '../GlobalStyles';

const Overlay = styled.div`
  background-color: ${GlobalStylesTheme.black};
  bottom: 0;
  left: 0;
  opacity: .5;
  position: fixed;
  right: 0;
  top: 0;
  z-index: ${GlobalStylesTheme.zIndexBackdrop};
`;

// TODO: Add listener for escape key
function Backdrop ({children, show}) {
  function renderBackdrop () {
    return (
      <Overlay>{children}</Overlay>
    );
  }

  return (
    <Fragment>{show && renderBackdrop()}</Fragment>
  );
}

Backdrop.propTypes = {
  children: PropTypes.element,
  show: PropTypes.bool.isRequired
};

Backdrop.defaultProps = {
  children: null
};

export default memo(Backdrop);
