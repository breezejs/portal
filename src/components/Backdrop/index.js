export {default as default} from './Backdrop';
