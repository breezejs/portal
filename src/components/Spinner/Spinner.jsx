import {darken} from 'polished';
import PropTypes from 'prop-types';
import React, {memo} from 'react';
import styled from 'styled-components';
import {GlobalStylesTheme} from '../GlobalStyles';

const SpinnerElement = styled.div`
  @keyframes spinner {
    to {
      transform: rotate(360deg);
    }
    
  }

  height: 100px;
  left: 50%;
  margin: -50px 0 0 -50px;
  position: fixed;
  top: 50%;
  width: 100px;
  z-index: ${GlobalStylesTheme.zIndexSpinner};

  ::before {
    content: "Loading...";
    height: 100px;
    left: 50%;
    margin: -50px 0 0 -50px;
    position: absolute;
    top: 50%;
    width: 100px;
  }

  :not(:required) {
    ::before {
      animation: spinner .6s linear infinite;
      border: 5px solid ${darken(0.1, GlobalStylesTheme.bodyBg)};
      border-radius: 50%;
      border-top-color: ${({theme}) => GlobalStylesTheme[theme]};
      content: "";
    }
  }
`;

function Spinner ({theme}) {
  return (
    <SpinnerElement theme={theme} />
  );
}

Spinner.propTypes = {
  theme: PropTypes.oneOf(['danger', 'info', 'primary', 'secondary', 'success', 'warning'])
};

export default memo(Spinner);
