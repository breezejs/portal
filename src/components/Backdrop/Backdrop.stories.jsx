import {boolean, withKnobs} from '@storybook/addon-knobs';
import {storiesOf} from '@storybook/react';
import React from 'react';
import GlobalStyles from '../GlobalStyles';
import Spinner from '../Spinner';
import Backdrop from './Backdrop';

const stories = storiesOf('Backdrop', module);

stories
  .addDecorator(withKnobs)
  .addDecorator((storyFn) => (
    <div>
      <GlobalStyles />
      {storyFn()}
    </div>
  ))
  .add('with no content', () => (
    <Backdrop show={boolean('Show', true)} />
  ))
  .add('with spinner', () => (
    <Backdrop show={boolean('Show', true)}>
      <Spinner theme="primary" />
    </Backdrop>
  ));
