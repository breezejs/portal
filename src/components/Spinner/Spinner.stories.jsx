import {optionsKnob, withKnobs} from '@storybook/addon-knobs';
import {storiesOf} from '@storybook/react';
import React from 'react';
import GlobalStyles from '../GlobalStyles';
import Spinner from './Spinner';

const stories = storiesOf('Spinner', module);

const options = {
  danger: 'danger',
  info: 'info',
  primary: 'primary',
  secondary: 'secondary',
  success: 'success',
  warning: 'warning'
};
const optionsObj = {
  display: 'select'
};

stories
  .addDecorator(withKnobs)
  .addDecorator((storyFn) => (
    <div>
      <GlobalStyles />
      {storyFn()}
    </div>
  ))
  .add('with multiple themes', () => (
    <Spinner theme={optionsKnob('Theme', options, 'info', optionsObj)} />
  ));
