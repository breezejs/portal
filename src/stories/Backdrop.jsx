import {boolean, withKnobs} from '@storybook/addon-knobs';
import {storiesOf} from '@storybook/react';
import React from 'react';
import Backdrop from '../components/Backdrop';
import GlobalStyles from '../components/GlobalStyles';
import Spinner from '../components/Spinner';

const stories = storiesOf('Backdrop', module);

stories
  .addDecorator(withKnobs)
  .addDecorator((storyFn) => (
    <div>
      <GlobalStyles />
      {storyFn()}
    </div>
  ))
  .add('with no content', () => (
    <Backdrop show={boolean('Show', true)} />
  ))
  .add('with spinner', () => (
    <Backdrop show={boolean('Show', true)}>
      <Spinner theme="primary" />
    </Backdrop>
  ));
