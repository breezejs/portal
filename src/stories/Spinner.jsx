import {storiesOf} from '@storybook/react';
import React from 'react';
import GlobalStyles from '../components/GlobalStyles';
import Spinner from '../components/Spinner';

const stories = storiesOf('Spinner', module);

stories
  .addDecorator((storyFn) => (
    <div>
      <GlobalStyles />
      {storyFn()}
    </div>
  ))
  .add('with primary theme', () => (
    <Spinner theme="primary" />
  ))
  .add('with secondary theme', () => (
    <Spinner theme="secondary" />
  ))
  .add('with info theme', () => (
    <Spinner theme="info" />
  ))
  .add('with success theme', () => (
    <Spinner theme="success" />
  ))
  .add('with danger theme', () => (
    <Spinner theme="danger" />
  ))
  .add('with warning theme', () => (
    <Spinner theme="warning" />
  ));
