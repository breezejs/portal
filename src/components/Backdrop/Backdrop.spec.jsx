import {mount, shallow} from 'enzyme';
import React from 'react';
import Backdrop from './Backdrop';
import Spinner from '../Spinner';

describe('Backdrop', () => {
  describe('with children', () => {
    it('should render when shown', () => {
      const result = mount(
        <Backdrop show={true}>
          <Spinner />
        </Backdrop>
      );

      expect(result).toMatchSnapshot();
    });

    it('should render when not shown', () => {
      const result = shallow(
        <Backdrop show={false}>
          <Spinner />
        </Backdrop>
      );

      expect(result).toMatchSnapshot();
    });
  });

  describe('without children', () => {
    it('should render when shown', () => {
      const result = shallow(
        <Backdrop show={true} />
      );

      expect(result).toMatchSnapshot();
    });

    it('should render when not shown', () => {
      const result = shallow(
        <Backdrop show={false} />
      );

      expect(result).toMatchSnapshot();
    });
  });
});
