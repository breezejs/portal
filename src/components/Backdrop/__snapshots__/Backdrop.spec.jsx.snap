// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Backdrop with children should render when not shown 1`] = `<Fragment />`;

exports[`Backdrop with children should render when shown 1`] = `
.c0 {
  background-color: #000;
  bottom: 0;
  left: 0;
  opacity: .5;
  position: fixed;
  right: 0;
  top: 0;
  z-index: 1000;
}

.c1 {
  height: 100px;
  left: 50%;
  margin: -50px 0 0 -50px;
  position: fixed;
  top: 50%;
  width: 100px;
  z-index: 1100;
}

.c1::before {
  content: "Loading...";
  height: 100px;
  left: 50%;
  margin: -50px 0 0 -50px;
  position: absolute;
  top: 50%;
  width: 100px;
}

.c1:not(:required)::before {
  -webkit-animation: spinner .6s linear infinite;
  animation: spinner .6s linear infinite;
  border: 5px solid #e6e6e6;
  border-radius: 50%;
  content: "";
}

<Backdrop
  show={true}
>
  <StyledComponent
    forwardedComponent={
      Object {
        "$$typeof": Symbol(react.forward_ref),
        "attrs": Array [],
        "componentStyle": ComponentStyle {
          "componentId": "sc-bdVaJa",
          "isStatic": true,
          "lastClassName": "c0",
          "rules": Array [
            "
  background-color: ",
            "#000",
            ";
  bottom: 0;
  left: 0;
  opacity: .5;
  position: fixed;
  right: 0;
  top: 0;
  z-index: ",
            "1000",
            ";
",
          ],
        },
        "displayName": "styled.div",
        "foldedComponentIds": Array [],
        "render": [Function],
        "styledComponentId": "sc-bdVaJa",
        "target": "div",
        "toString": [Function],
        "warnTooManyClasses": [Function],
        "withComponent": [Function],
      }
    }
    forwardedRef={null}
  >
    <div
      className="c0"
    >
      <Spinner>
        <StyledComponent
          forwardedComponent={
            Object {
              "$$typeof": Symbol(react.forward_ref),
              "attrs": Array [],
              "componentStyle": ComponentStyle {
                "componentId": "sc-bwzfXH",
                "isStatic": false,
                "lastClassName": "c1",
                "rules": Array [
                  "
  @keyframes spinner {
    to {
      transform: rotate(360deg);
    }
    
  }

  height: 100px;
  left: 50%;
  margin: -50px 0 0 -50px;
  position: fixed;
  top: 50%;
  width: 100px;
  z-index: ",
                  "1100",
                  ";

  ::before {
    content: \\"Loading...\\";
    height: 100px;
    left: 50%;
    margin: -50px 0 0 -50px;
    position: absolute;
    top: 50%;
    width: 100px;
  }

  :not(:required) {
    ::before {
      animation: spinner .6s linear infinite;
      border: 5px solid ",
                  "#e6e6e6",
                  ";
      border-radius: 50%;
      border-top-color: ",
                  [Function],
                  ";
      content: \\"\\";
    }
  }
",
                ],
              },
              "displayName": "styled.div",
              "foldedComponentIds": Array [],
              "render": [Function],
              "styledComponentId": "sc-bwzfXH",
              "target": "div",
              "toString": [Function],
              "warnTooManyClasses": [Function],
              "withComponent": [Function],
            }
          }
          forwardedRef={null}
        >
          <div
            className="c1"
          />
        </StyledComponent>
      </Spinner>
    </div>
  </StyledComponent>
</Backdrop>
`;

exports[`Backdrop without children should render when not shown 1`] = `<Fragment />`;

exports[`Backdrop without children should render when shown 1`] = `
<Fragment>
  <styled.div />
</Fragment>
`;
